File: web/src/utils/sse.ts
/**
 * SSE 公共消费封装（对应 change add-streaming-pipeline 任务 1.2.1）。
 *
 * 用原生 fetch + ReadableStream 手写 SSE 解析（零依赖）：
 * - POST + Authorization 鉴权头（浏览器原生 EventSource 仅支持 GET 且不便带自定义头，不可用）
 * - 按 \n\n 切事件块，每块解析 event:/data: 行（注释行如心跳 : keepalive 忽略）
 * - 流式端点直连后端（VITE_STREAM_BASE_URL），绕开 vite dev proxy（proxy 不转发 text/event-stream）
 * - 返回 AbortController，调用方可随时取消
 *
 * 三模块（简历解析 / 简历优化 / JD 匹配）共用。
 */

/** SSE 事件回调：按 event 类型分发，只需提供关心的回调（其余可省略） */
export interface SseHandlers {
  /** 阶段进度 */
  onStage?: (p: { message: string; node?: string }) => void
  /** 分段中间结果（结构化任务，如画像的一段） */
  onSegment?: (p: { section: string; data: any; confidence?: any }) => void
  /** 增量文本内容（自由文本任务，如简历生成） */
  onToken?: (p: { delta: string }) => void
  /** AI 思考过程（glm-4.5 的 reasoning_content 思考 token） */
  onReasoning?: (p: { delta: string }) => void
  /** 评分（JD 匹配） */
  onScore?: (p: any) => void
  /** 任务完成（最终结果） */
  onDone?: (p: any) => void
  /** 增补完成（JD 匹配） */
  onEnriched?: (p: { gaps: any[] }) => void
  /** 失败 */
  onError?: (p: { error_code: string; error_message: string }) => void
}

/**
 * 打开一条 SSE 流：POST body 到 url，按 handlers 分发事件。
 *
 * @param url      API 路径（如 '/jd/match'，会拼到 VITE_STREAM_BASE_URL 后面）
 * @param body     请求体（POST，会被 JSON.stringify）
 * @param handlers 事件回调
 * @returns AbortController（调 .abort() 取消任务；取消时不触发 onError）
 */
export function openSseStream(
  url: string,
  body: unknown,
  handlers: SseHandlers,
): AbortController {
  const controller = new AbortController()

  void (async () => {
    // SSE 直连后端（VITE_STREAM_BASE_URL），绕开 vite proxy——proxy 不流式转发 text/event-stream。
    // 未配置时回退到普通 API base（走 proxy，流式可能不工作但至少不报错，便于降级）。
    const base = import.meta.env['VITE_STREAM_BASE_URL'] || import.meta.env.VITE_API_BASE_URL || ''
    const token = localStorage.getItem('token')
    const headers: Record<string, string> = { 'Content-Type': 'application/json' }
    if (token) headers['Authorization'] = `Bearer ${token}`

    let resp: Response
    try {
      resp = await fetch(`${base}${url}`, {
        method: 'POST',
        headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      })
    } catch (e: any) {
      if (e?.name === 'AbortError') return
      handlers.onError?.({ error_code: 'NETWORK', error_message: e?.message || '网络错误' })
      return
    }

    if (!resp.ok || !resp.body) {
      let message = `请求失败（${resp.status}）`
      try {
        const j = await resp.json()
        message = j?.message || message
      } catch {
        /* 非 JSON 错误体，保留默认文案 */
      }
      handlers.onError?.({ error_code: 'HTTP_' + resp.status, error_message: message })
      return
    }

    // 手写 SSE 解析：流按 \n\n 切事件块，每块解析 event:/data: 行
    const reader = resp.body.getReader()
    const decoder = new TextDecoder('utf-8')
    let buffer = ''
    try {
      while (true) {
        const { value, done } = await reader.read()
        if (done) break
        buffer += decoder.decode(value, { stream: true })
        let sep = buffer.indexOf('\n\n')
        while (sep !== -1) {
          const evt = parseSse(buffer.slice(0, sep))
          buffer = buffer.slice(sep + 2)
          if (evt) dispatch(evt, handlers)
          sep = buffer.indexOf('\n\n')
        }
      }
    } catch (e: any) {
      if (e?.name !== 'AbortError') {
        handlers.onError?.({ error_code: 'STREAM', error_message: e?.message || '流读取中断' })
      }
    }
  })()

  return controller
}

/** 解析单个 SSE 事件块（event:/data: 行；注释行如 : keepalive 忽略） */
function parseSse(raw: string): { event: string; data: any } | null {
  let event = 'message'
  let dataStr = ''
  for (const line of raw.split('\n')) {
    if (line.startsWith('event:')) event = line.slice(6).trim()
    else if (line.startsWith('data:')) dataStr += line.slice(5).trim()
    // 以 : 开头的注释行（保活心跳）忽略
  }
  if (!dataStr) return null
  try {
    return { event, data: JSON.parse(dataStr) }
  } catch {
    return null
  }
}

/** 按 event 类型分发到 handlers 对应回调 */
function dispatch(evt: { event: string; data: any }, h: SseHandlers) {
  switch (evt.event) {
    case 'stage': h.onStage?.(evt.data); break
    case 'segment': h.onSegment?.(evt.data); break
    case 'token': h.onToken?.(evt.data); break
    case 'reasoning': h.onReasoning?.(evt.data); break
    case 'score': h.onScore?.(evt.data); break
    case 'done': h.onDone?.(evt.data); break
    case 'enriched': h.onEnriched?.(evt.data); break
    case 'error': h.onError?.(evt.data); break
    default: break
  }
}

File: backend/src/core/sse.py
"""
SSE（Server-Sent Events）公共工具

为长耗时 LLM 任务的流式管线提供统一的事件组装能力（对应 change
add-streaming-pipeline）。三条链路（简历解析 / 简历优化 / JD 匹配）共用。

SSE 协议：每个事件由若干行组成，以空行（\\n\\n）结束：
    event: <事件名>\\n
    data: <JSON 字符串>\\n
    \\n
前端按 \\n\\n 切块、解析 event:/data: 行（见 web/src/utils/sse.ts）。

心跳（保活）：SSE 注释行（以 : 开头）不会触发前端事件，但会刷新连接、
防止 idle 超时——长任务某段无内容可推时定期 yield 它保活。

作者：求职 Copilot 项目
日期：2026-07-23
"""
import json
from typing import Any, Dict


# ============================================================================
# 事件类型常量（三模块共用，对应 design.md「事件协议」表）
# ============================================================================
STAGE = "stage"          # 阶段进度 {message, node?}
SEGMENT = "segment"      # 分段中间结果（结构化任务，如画像的一段）{section, data, confidence?}
TOKEN = "token"          # 增量文本内容（自由文本任务，如简历生成）{delta}
REASONING = "reasoning"  # AI 思考过程（glm-4.5 的 reasoning_content）{delta}
SCORE = "score"          # 评分（JD 匹配）{overall_score, dimension_scores, ...}
DONE = "done"            # 任务完成（最终结果）
ENRICHED = "enriched"    # 增补完成（JD 匹配）{gaps}
ERROR = "error"          # 失败 {error_code, error_message}


# ============================================================================
# 流式响应头
# ============================================================================
# Cache-Control: no-cache —— 禁止中间代理缓存流式响应
# X-Accel-Buffering: no  —— 禁用 nginx 缓冲（若有反代），确保事件即时下发
STREAM_HEADERS = {"Cache-Control": "no-cache", "X-Accel-Buffering": "no"}
STREAM_MEDIA_TYPE = "text/event-stream"


def sse(event: str, data: Dict[str, Any]) -> str:
    """
    组装一条 SSE 事件字符串。

    Args:
        event: 事件名（用上面的常量）
        data:  事件数据（会被 JSON 序列化，ensure_ascii=False 保留中文）
    Returns:
        形如 "event: stage\\ndata: {...}\\n\\n" 的字符串，可直接 yield

    示例：
        >>> sse("stage", {"message": "正在解析…"})
        'event: stage\\ndata: {"message": "正在解析…"}\\n\\n'
    """
    return f"event: {event}\ndata: {json.dumps(data, ensure_ascii=False)}\n\n"


def sse_keepalive() -> str:
    """
    保活心跳：SSE 注释行（前端忽略，但连接保持活跃）。

    长任务执行中若某段暂时无内容可推（如单段 LLM 仍在生成、未到整段下发时机），
    定期 yield 本函数返回值，防止前端 / 代理 / 浏览器因 idle 而切断连接。

    Returns:
        ': keepalive\\n\\n'
    """
    return ": keepalive\n\n"
